### PR TITLE
Allow calls to IsManagedType while computing GetMembers

### LINKED ISSUE #10195: StackOverflowException compiling partial method with struct* parameter
Compile the following with ```csc -t:library c.cs```

```
unsafe partial struct S
{
    partial void M(S* p) { }
}
```

### PR DESCRIPTION
**Pull request by cston**
Fixes #10195

### CODE REVIEW

**Comment by cston** [comment]
@dotnet/roslyn-compiler please review.

File: src/Compilers/CSharp/Portable/Binder/BinderFactory.cs [hunk 1]
```diff
@@ -89,31 +89,31 @@ private bool InScript
         /// <summary>
         /// Note, there is no guarantee that the factory always gives back the same binder instance for the same <param name="node"/>.
         /// </summary>
-        internal Binder GetBinder(CSharpSyntaxNode node)
+        /// <param name="memberDeclarationOpt"/>
+        /// <param name="memberOpt"/>
+        internal Binder GetBinder(CSharpSyntaxNode node, CSharpSyntaxNode memberDeclarationOpt = null, Symbol memberOpt = null)
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/BinderFactory.cs]: Can we get a comment as to the intent of these parameters?

Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/BinderFactory.cs]: Update comment.

File: src/Compilers/CSharp/Portable/Binder/BinderFactory.cs [hunk 2]
```diff
@@ -87,34 +87,40 @@ private bool InScript
...
+        /// than looking up the member in the containing type, allowing this method to be called
+        /// while calculating the member list.
         /// </summary>
-        internal Binder GetBinder(CSharpSyntaxNode node)
+        /// <remarks>
+        /// Note, there is no guarantee that the factory always gives back the same binder instance for the same node.
+        /// </remarks>
+        internal Binder GetBinder(CSharpSyntaxNode node, CSharpSyntaxNode memberDeclarationOpt = null, Symbol memberOpt = null)
         {
+            Debug.Assert((memberDeclarationOpt == null) == (memberOpt == null));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/BinderFactory.cs]: This assert probably belongs in BinderFactoryVisitor.Initialize method.

File: src/Compilers/CSharp/Portable/Symbols/BaseTypeAnalysis.cs [hunk 3]
```diff
@@ -1,87 +1,69 @@
...
-        internal static bool ClassDependsOn(TypeSymbol depends, TypeSymbol on)
+        internal static bool ClassDependsOn(NamedTypeSymbol depends, NamedTypeSymbol on)
         {
-            if ((object)depends == null || (object)on == null)
-            {
-                return false;
-            }
+            Debug.Assert((object)depends != null);
+            Debug.Assert((object)on != null);
+            Debug.Assert((object)on.OriginalDefinition == on);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/BaseTypeAnalysis.cs]: on.IsDefinition?

File: src/Compilers/CSharp/Portable/Symbols/BaseTypeAnalysis.cs [hunk 4]
```diff
@@ -1,87 +1,69 @@
...
         {
-            if ((object)type == null)
-            {
-                return;
-            }
+            Debug.Assert((object)type != null);
 
-            var nt = type as NamedTypeSymbol;
-            if ((object)nt != null && ReferenceEquals(nt.OriginalDefinition, on))
+            if (ReferenceEquals(type.OriginalDefinition, on))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/BaseTypeAnalysis.cs]: ```(object)type.OriginalDefinition == on```?

File: src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs [hunk 5]
```diff
@@ -562,6 +562,32 @@ internal virtual ImmutableArray<Symbol> GetSimpleNonTypeMembers(string name)
...
         /// <summary>
+        /// Get all instance field and event members.
+        /// </summary>
+        /// <remarks>
+        /// For source symbols may be called while calculating
+        /// <see cref="NamespaceOrTypeSymbol.GetMembersUnordered"/>.
+        /// </remarks>
+        internal virtual IEnumerable<Symbol> GetInstanceFieldsAndEvents()
+        {
+            return GetMembersUnordered().Where(IsInstanceFieldOrEvent);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs]: We should probably cache this delegate.

File: src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingNamedTypeSymbol.cs [hunk 6]
```diff
@@ -445,6 +441,10 @@ internal override NamedTypeSymbol BaseTypeNoUseSiteDiagnostics
                             acyclicBase = this.RetargetingTranslator.Retarget(underlyingBase, RetargetOptions.RetargetPrimitiveTypesByName);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingNamedTypeSymbol.cs]: Should this be a subject of the cyclic inheritance check?

File: src/Compilers/CSharp/Portable/Binder/BinderFactory.cs (hunk 2 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/BinderFactory.cs]: Moved assert.

File: src/Compilers/CSharp/Portable/Symbols/BaseTypeAnalysis.cs (hunk 3 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/BaseTypeAnalysis.cs]: Changed.

File: src/Compilers/CSharp/Portable/Symbols/BaseTypeAnalysis.cs (hunk 4 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/BaseTypeAnalysis.cs]: Changed.

File: src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs (hunk 5 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs]: Done.

File: src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingNamedTypeSymbol.cs (hunk 6 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingNamedTypeSymbol.cs]: Will this base type be a type other than `System.Object` or `System.ValueType`?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingNamedTypeSymbol.cs]: I am not sure. Is there any harm to do the check in this case?

Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingNamedTypeSymbol.cs]: Updated.

**Comment by cston** [comment]
@alekseyts, @jaredpar all feedback has been addressed.

**Comment by cston** [comment]
@dotnet/roslyn-compiler please review.

**Comment by jaredpar** [comment]
:+1:

**Comment by AlekseyTs** [comment]
LGTM

**Pull request merged**